messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupAdd.java

```diff
@@ -19,132 +19,73 @@
+public class DiscoveryGroupAdd extends ShallowResourceAdd {
 
-    private DiscoveryGroupAdd() {
-        super(DiscoveryGroupDefinition.ATTRIBUTES);
-    }
+    public static final DiscoveryGroupAdd INSTANCE = new DiscoveryGroupAdd(false);
+    public static final DiscoveryGroupAdd LEGACY_INSTANCE = new DiscoveryGroupAdd(true);
 
-    private static boolean isSubsystemResource(final OperationContext context) {
-        return ModelDescriptionConstants.SUBSYSTEM.equals(context.getCurrentAddress().getParent().getLastElement().getKey());
+    private DiscoveryGroupAdd(boolean isLegacyCall) {
+        super(isLegacyCall, DiscoveryGroupDefinition.ATTRIBUTES);
     }
 
     @Override
     public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
         CommonAttributes.renameChannelToCluster(operation);
-        if (operation.hasDefined(JGROUPS_CLUSTER.getName())) {
-            if (operation.hasDefined(JGROUPS_CHANNEL_FACTORY.getName()) && !operation.hasDefined(JGROUPS_CHANNEL.getName())) {
-                // Handle legacy behavior
-                String channel = operation.get(JGROUPS_CLUSTER.getName()).asString();
-                operation.get(JGROUPS_CHANNEL.getName()).set(channel);
-
-
-                final PathAddress channelAddress;
-                if (isSubsystemResource(context)) {
-                    channelAddress = context.getCurrentAddress().getParent().getParent().append(ModelDescriptionConstants.SUBSYSTEM, "jgroups").append("channel", channel);
-                } else {
-                    channelAddress = context.getCurrentAddress().getParent().getParent().getParent().append(ModelDescriptionConstants.SUBSYSTEM, "jgroups").append("channel", channel);
-                }
-                ModelNode addChannelOperation = Util.createAddOperation(channelAddress);
-                addChannelOperation.get("stack").set(operation.get(JGROUPS_CHANNEL_FACTORY.getName()));
-                // Fabricate a channel resource
-                context.addStep(addChannelOperation, AddIfAbsentStepHandler.INSTANCE, OperationContext.Stage.MODEL);
+        if (!isLegacyCall) {
+            ModelNode op = operation.clone();
+            PathAddress target = context.getCurrentAddress().getParent();
+            OperationStepHandler addHandler;
+            if (operation.hasDefined(JGROUPS_CLUSTER.getName())) {
+                target = target.append(CommonAttributes.JGROUP_DISCOVERY_GROUP, context.getCurrentAddressValue());
+                addHandler = JGroupDiscoveryGroupAdd.LEGACY_INSTANCE;
+            } else {
+                target = target.append(CommonAttributes.SOCKET_DISCOVERY_GROUP, context.getCurrentAddressValue());
+                addHandler = SocketDiscoveryGroupAdd.LEGACY_INSTANCE;
             }
+            op.get(OP_ADDR).set(target.toModelNode());
+            context.addStep(op, addHandler, OperationContext.Stage.MODEL, true);
```

Comment:
Why use an intermediate step?  Can't we convert these operations to use their non-legacy versions here?